control/state.py

```diff
@@ -163,156 +272,214 @@ def _remove_key(self, key: str):
-                req.nsid = int(nsid)
-                callback(req)
+    Instance attributes:
+        config: Basic gateway parameters
+        logger: Logger instance to track events
+        local: Local GatewayState instance
+        remote: Remote GatewayState instance
+        update_interval: Interval to periodically poll for updates
+        update_timer: Timer to check for gateway state updates
+    """
 
-    def add_subsystem(self, subsystem_nqn: str, val: str):
-        """Adds a subsystem to the OMAP."""
-        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
-        self._add_key(key, val)
+    def __init__(self, config, local, remote):
+        self.config = config
+        self.local = local
+        self.remote = remote
+        self.logger = logging.getLogger(__name__)
 
-    def remove_subsystem(self, subsystem_nqn: str):
-        """Removes a subsystem from the OMAP."""
-        key = self.SUBSYSTEM_PREFIX + subsystem_nqn
-        self._remove_key(key)
+    def add_bdev(self, bdev_name: str, val: str):
+        """Adds a bdev to the state data stores."""
+        self.remote.add_bdev(bdev_name, val)
+        self.local.add_bdev(bdev_name, val)
 
-        # Delete all keys related to subsystem
-        omap_dict = self._read_all()
-        for key in omap_dict.keys():
-            if (key.startswith(self.NAMESPACE_PREFIX + subsystem_nqn) or
-                    key.startswith(self.HOST_PREFIX + subsystem_nqn) or
-                    key.startswith(self.LISTENER_PREFIX + subsystem_nqn)):
-                self._remove_key(key)
+    def remove_bdev(self, bdev_name: str):
+        """Removes a bdev from the state data stores."""
+        self.remote.remove_bdev(bdev_name)
+        self.local.remove_bdev(bdev_name)
 
-    def _restore_subsystems(self, omap_dict, callback):
-        """Restores subsystems from the OMAP."""
+    def add_namespace(self, subsystem_nqn: str, nsid: str, val: str):
+        """Adds a namespace to the state data store."""
+        self.remote.add_namespace(subsystem_nqn, nsid, val)
+        self.local.add_namespace(subsystem_nqn, nsid, val)
 
-        for (key, val) in omap_dict.items():
-            if key.startswith(self.SUBSYSTEM_PREFIX):
-                req = json_format.Parse(val, pb2.create_subsystem_req())
-                callback(req)
+    def remove_namespace(self, subsystem_nqn: str, nsid: str):
+        """Removes a namespace from the state data store."""
+        self.remote.remove_namespace(subsystem_nqn, nsid)
+        self.local.remove_namespace(subsystem_nqn, nsid)
 
-    def add_host(self, subsystem_nqn: str, host_nqn: str, val: str):
-        """Adds a host to the OMAP."""
-        key = "{}{}_{}".format(self.HOST_PREFIX, subsystem_nqn, host_nqn)
-        self._add_key(key, val)
+    def add_subsystem(self, subsystem_nqn: str, val: str):
+        """Adds a subsystem to the state data store."""
+        self.remote.add_subsystem(subsystem_nqn, val)
```

Comment:
If this fails because our local version is not up to date should we update and retry here or elsewhere? What if someone else added the same subsystem in the meantime. We already configured SPDK at this point so this is probably not going to work. Otherwise we would have to revert SPDK changes.

Reply:
Probably need to change the order to first update the state then update SPDK in case it fails revert the state change.